RFD.md

```diff
@@ -0,0 +1,267 @@
+
+### CLI
+
+Users will be able to run a CLI program that will manage jobs on a remote Linux system. This CLI tool will be called `rps` and will have the following commands:
+
+- `rps start -c "<command>"`: Start a new job with the given command and return the job_id
+- `rps status <job_id>`: Get the status of the job with the given job_id
+- `rps output <job_id>`: Stream the output of the job with the given job_id
+- `rps stop <job_id>`: Stop a running job with the given job_id
+
+The hostname/IP address will be hard-coded for simplicity
+
+`job_id` is a unique identifier for a job, implemented as a UUID.
+
+Here are some examples of how the CLI will work:
+
+```sh
+$ rps start -c "for i in {1..10}; do echo $i; sleep 1; done"
+Job started with job_id: 12345678-1234-1234-1234-123456789012
+$ rps status 12345678-1234-1234-1234-123456789012
+Job 12345678-1234-1234-1234-123456789012 is running
+- started: 2025-04-20 12:31:45 -0500 CDT
+- command: "for i in {1..10}; do echo $i; sleep 1; done"
+$ rps output 12345678-1234-1234-1234-123456789012
+1
+2
+3
+4
+$ rps stop 12345678-1234-1234-1234-123456789012
+Job 12345678-1234-1234-1234-123456789012 is stopped
+```
+
+### API
+
+#### Authentication
+
+Authentication will be handled using mTLS. TLS 1.3 with `TLS_AES_256_GCM_SHA384` cipher suite will be used for maximum security.
+
+For simplicity, the certificates (CA, server, client) will be generated using `openssl` with the `Ed25519` algorithm and checked into source. In a real application, the certificates should be generated and stored in a secure manner.
+
+#### Authorization
+
+The server will use the client certificate's `Subject.CommonName` as the `clientId` to identify the client.
+
+Authorization will be segmented into:
+
+- `rpsreader`: read-only access (`GetJobStatus`, `StreamJobOutput`)
+- `rpsmanager`: full job management access (read + `StartJob` & `StopJob`)
+
+Users (clients) will be assigned one of these two roles in a hard-coded map for simplicity. In a real application, this assignment would be handled by an authorization server with groups and permissions.
+
+#### GRPC
+
+The API will be a gRPC server that handles messages from the CLI. Protocol Buffers (protobuf) will be used to define and serialize the methods and messages, which will provide consistency between the gRPC server and client.
+
+The following proto interface will be used:
+```proto
+// Service definition for the Remote Process Service (RPS)
+service RpsService {
+  // Starts a new job with the given command.
+  rpc Start(StartRequest) returns (StartResponse);
+
+  // Gets the status of a specific job.
+  rpc Status(StatusRequest) returns (StatusResponse);
+
+  // Streams the output of a specific job.
+  rpc Output(OutputRequest) returns (stream OutputResponse);
+
+  // Stops a running job.
+  rpc Stop(StopRequest) returns (StopResponse);
+}
+
+// Request message for starting a job.
+message StartRequest {
+  string command = 1; // The command to execute for the job.
```

Comment:
Does the command’s arguments also get forwarded via the command field?

Does  it make sense to extend the message to include a repeated string args field, so that the command and its arguments can be structured separately?